### PR TITLE
python: Rewrite path injection query to use flow state

### PR DESCRIPTION
**Pull request by yoff**
This removes the FP caused by chaining.

This PR also removes `ChainedConfigs12.qll`, as we hope to solve future problems via flow states.

I have marked it as not needing a change note, but let me know if you think it does :-)

### CODE REVIEW

**Review by RasmusWL — CHANGES_REQUESTED** [review]
Very cool that you have porting these queries to use flow state :muscle: (also great to see that FP removed)

Is there an official doc for how to use flow state? (so I know if you made up the way to handle things, or if there is an established way already)

For `python/ql/lib/semmle/python/security/dataflow/PathInjection.qll`, I'm wondering how much deprecation we need to support. At least if someone has code that uses the old configurations (or old predicate), their code is going to stop working if this PR is merged as it is right now.

I also have a minor comment for the chained-config qll file, where I think we can mark even more things as deprecated.

File: python/ql/lib/semmle/python/security/dataflow/ChainedConfigs12.qll [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+ */
+private newtype TCustomPathNode =
+  Config1Node(DataFlow::PathNode node1) { not crossoverNode(node1.getNode()) } or
+  Config2Node(DataFlow2::PathNode node2) { not crossoverNode(node2.getNode()) } or
+  CrossoverNode(DataFlow::Node node) { crossoverNode(node) }
+
+/**
+ * A class representing the set of all the path nodes in either config.
+ */
+class CustomPathNode extends TCustomPathNode {
```
[the remark this answers is not in the document]
Comment (yoff) [reply python/ql/lib/semmle/python/security/dataflow/ChainedConfigs12.qll:35]: I have nothing against this, but I thought I had deprecated the whole file?

File: python/ql/src/experimental/semmle/python/security/injection/NoSQLInjection.qll [hunk 2]
```diff
@@ -1,57 +1,54 @@
...
+    override predicate isSink(DataFlow::Node sink, DataFlow::FlowState state) {
+      sink = any(NoSQLQuery noSQLQuery).getQuery() and
+      state instanceof ConvertedToDict
+    }
+
+    override predicate isBarrier(DataFlow::Node node, DataFlow::FlowState state) {
+      // Block `RemoteInput` paths here, since they change state to `ConvertedToDict`
+      exists(Decoding decoding | decoding.getFormat() = "JSON" and node = decoding.getOutput()) and
+      state instanceof RemoteInput
+    }
```
[the remark this answers is not in the document]
Comment (yoff) [reply python/ql/src/experimental/semmle/python/security/injection/NoSQLInjection.qll:26]: We are only blocking flow on the output, so I believe both instances of `user_input` will be `"RemoteInput"` while both `decoded_1` and `decoded_2` will be `"ConvertedToDict"`.

**Comment by yoff** [comment]
> Is there an official doc for how to use flow state? (so I know if you made up the way to handle things, or if there is an established way already)

I am not aware of any official doc, so this is all made up, but it seemed pretty intuitive :-)
(I am happy for best practices to evolve, though. JavaScript has a fairly advanced flow-state based query for path injection that we should probably attempt to port at some point.)

**Comment by yoff** [comment]
Good point about the old configuration being public. We may have to leave it deprecated instead of deleting it.

File: python/ql/lib/semmle/python/security/dataflow/ChainedConfigs12.qll (hunk 1 above)
[the remark this answers is not in the document]
Comment (yoff) [reply python/ql/lib/semmle/python/security/dataflow/ChainedConfigs12.qll]: I see, there is a formal `detracted` keyword. That will help discover the deprecation. I also deprecated the two predicates.

File: python/ql/src/experimental/semmle/python/security/injection/NoSQLInjection.qll (hunk 2 above)
[the remark this answers is not in the document]
Comment (RasmusWL) [reply python/ql/src/experimental/semmle/python/security/injection/NoSQLInjection.qll:26]: yes, I overlooked the part about _flow on the output_. Thanks :+1:

**Review by RasmusWL — APPROVED** [review]
:+1: with the deprecations.

I'm still a bit surprised that the flow-state specific sanitizer is `isBarrier/2`, but :shrug: 

Out of interest, did you come up with the scheme of blocking flow for the old state when there is a state transition that occurs? (as with the barrier for NoSQL injection) I guess it helps with performance on the data-flow library, so I can see the benefit for sure :+1:

**Comment by yoff** [comment]
> Out of interest, did you come up with the scheme of blocking flow for the old state when there is a state transition that occurs?

Yes. I guess we should get the conversation going about how to actually use flow states effectively, but in many cases it has to be blocked or it will reach the sink.

**Comment by yoff** [comment]
> I'm still a bit surprised that the flow-state specific sanitizer is `isBarrier/2`, but

I think it is because it is implemented for the data flow configurations, and no taint-specific veneer has been added to taint configurations...

**Pull request merged**